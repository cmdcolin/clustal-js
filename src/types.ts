declare interface Header {
  version: string;
  info: string;
}
declare interface Alignment {
  id: string;
  seq: string;
}

declare interface Block {
  consensus: string;
  seqs: string[];
  ids: string[];
}
declare interface Results {
  header: Header;
  alns: Alignment[];
  consensus: string;
}
