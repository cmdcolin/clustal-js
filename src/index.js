import LocalFile from './localFile'

export default class ClustalParser {
  constructor({ filename }) {
    if (filename) {
      this.clustal = new LocalFile(filename)
    }
  }

  async parse() {
    const contents = (await this.clustal.readFile()).toString().split('\n')
    const arr = contents[Symbol.iterator]()
    let line = arr.next()
    if (line.done) throw new Error('empty file')
    else line = line.value

    const knownHeaders = [
      'CLUSTAL O',
      'CLUSTAL W',
      'CLUSTALW',
      'CLUSTALO',
      'CLUSTAL',
      'PROBCONS',
      'MUSCLE',
    ]
    const header = knownHeaders.find(l => line.startsWith(l))
    const rest = line.slice(header.length)
    if (header === undefined) {
      throw new Error(
        `${header} is not a known CLUSTAL header: ${knownHeaders.join(',')}`,
      )
    }

    let version
    const words = rest.split(' ')
    for (let word of words) {
      if (word[0] === '(' && word[word.length-1] === ')') {
        word = word.substring(1, word.length - 1)
        console.log(word)
      }
      if ('0123456789'.includes(word[0])) {
        version = word
        break
      }
    }
    console.log(header, version)
    //There should be two blank lines after the header line
    line = arr.next().value
    while(line.trim() === "") {
        line = arr.next().value
    }

    const ids = []
    const seqs = []
    const consensus = ""
    let seq_cols = null

    //Use the first block to get the sequence identifiers
    while(true) {
        if(line[0] !== " " && line.trim() !== "") {
            //Sequences identifier...
            const fields = line.trimEnd().split(' ')
            console.log(fields)

            //We expect there to be two fields, there can be an optional
            //"sequence number" field containing the letter count.
            if(fields.length !== 2)
                throw new Error("Could not parse line:\n"+line)

            ids.append(fields[0])
            seqs.append(fields[1])

            //Record the sequence position to get the consensus
            if(seq_cols === null) {
                start = fields[0].length + line.slice(fields[0].length).find(fields[1])
                end = start + fields[1].length
                seq_cols = [start, end]
            }
            console.log(fields[1], line.slice(start, end))
        }//remove
    }//remove

//             if len(fields) == 3:
//                 #This MAY be an old style file with a letter count...
//                 try:
//                     letters = int(fields[2])
//                 except ValueError:
//                     raise ValueError("Could not parse line, bad sequence number:\n%s" % line)
//                 if len(fields[1].replace("-","")) != letters:
//                     raise ValueError("Could not parse line, invalid sequence number:\n%s" % line)
//         elif line[0] == " ":
//             #Sequence consensus line...
//             assert len(ids) == len(seqs)
//             assert len(ids) > 0
//             assert seq_cols is not None
//             consensus = line[seq_cols]
//             assert not line[:seq_cols.start].strip()
//             assert not line[seq_cols.stop:].strip()
//             #Check for blank line (or end of file)
//             line = handle.readline()
//             assert line.strip() == ""
//             break
//         else:
//             #No consensus
//             break
//         line = handle.readline()
//         if not line : break #end of file

    // assert line.strip() == ""
    // assert seq_cols is not None

    // #Confirm all same length
    // for s in seqs:
    //     assert len(s) == len(seqs[0])
    // if consensus:
    //     assert len(consensus) == len(seqs[0])

    // #Loop over any remaining blocks...
    // done = False
    // while not done:
    //     #There should be a blank line between each block.
    //     #Also want to ignore any consensus line from the
    //     #previous block.
    //     while (not line) or line.strip() == "":
    //         line = handle.readline()
    //         if not line : break # end of file
    //     if not line : break # end of file

    //     if line.split(None,1)[0] in known_headers:
    //         #Found concatenated alignment.
    //         done = True
    //         self._header = line
    //         break

    //     for i in range(len(ids)):
    //         assert line[0] != " ", "Unexpected line:\n%s" % repr(line)
    //         fields = line.rstrip().split()

    //         #We expect there to be two fields, there can be an optional
    //         #"sequence number" field containing the letter count.
    //         if len(fields) < 2 or len(fields) > 3:
    //             raise ValueError("Could not parse line:\n%s" % repr(line))

    //         if fields[0] != ids[i]:
    //             raise ValueError("Identifiers out of order? Got '%s' but expected '%s'" \
    //                               % (fields[0], ids[i]))

    //         if fields[1] != line[seq_cols]:
    //             start = len(fields[0]) + line[len(fields[0]):].find(fields[1])
    //             assert start == seq_cols.start, 'Old location %s -> %i:XX' % (seq_cols, start)
    //             end = start + len(fields[1])
    //             seq_cols = slice(start, end)
    //             del start, end

    //         #Append the sequence
    //         seqs[i] += fields[1]
    //         assert len(seqs[i]) == len(seqs[0])

    //         if len(fields) == 3:
    //             #This MAY be an old style file with a letter count...
    //             try:
    //                 letters = int(fields[2])
    //             except ValueError:
    //                 raise ValueError("Could not parse line, bad sequence number:\n%s" % line)
    //             if len(seqs[i].replace("-","")) != letters:
    //                 raise ValueError("Could not parse line, invalid sequence number:\n%s" % line)

    //         #Read in the next line
    //         line = handle.readline()
    //     #There should now be a consensus line
    //     if consensus:
    //         assert line[0] == " "
    //         assert seq_cols is not None
    //         consensus += line[seq_cols]
    //         assert len(consensus) == len(seqs[0])
    //         assert not line[:seq_cols.start].strip()
    //         assert not line[seq_cols.stop:].strip()
    //         #Read in the next line
    //         line = handle.readline()

    // assert len(ids) == len(seqs)
    // if len(seqs) == 0 or len(seqs[0]) == 0:
    //     raise StopIteration

    // if self.records_per_alignment is not None \
    // and self.records_per_alignment != len(ids):
    //     raise ValueError("Found %i records in this alignment, told to expect %i" \
    //                      % (len(ids), self.records_per_alignment))

    // records = (SeqRecord(Seq(s, self.alphabet), id=i, description=i) \
    //            for (i,s) in zip(ids, seqs))
    // alignment = MultipleSeqAlignment(records, self.alphabet)
    // #TODO - Handle alignment annotation better, for now
    // #mimic the old parser in Bio.Clustalw
    // if version:
    //     alignment._version = version
    // if consensus:
    //     alignment_length = len(seqs[0])
    //     assert len(consensus) == alignment_length, \
    //            "Alignment length is %i, consensus length is %i, '%s'" \
  }
}
