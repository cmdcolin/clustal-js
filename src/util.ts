export function parseVersion(line: string): string {
  const res = line.match(/\(?(\d+(\.\d+)+)\)?/);
  return res && res.length > 1 ? res[1] : "";
}

export function getFirstNonEmptyLine(arr: Symbol.iterator): Results {
  // There should be two blank lines after the header line
  let line = arr.next().value;
  while (line !== undefined && line.trim() === "") {
    line = arr.next().value;
  }
  return line;
}
