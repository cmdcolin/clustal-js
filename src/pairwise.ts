import { getFirstNonEmptyLine } from './util'

export function getSeqBounds(line: string) {
  const fields = line.split(/\s+/)
  const k1 = fields[0].length + fields[1].length
  const temp = line.slice(k1)
  const s = k1 + temp.indexOf(fields[2])
  const e = s + fields[2].length
  return [s, e] as const
}

// Use the first block to get the sequence identifiers
export function parsePairwiseBlock(arr: Iterator<string>) {
  let line = getFirstNonEmptyLine(arr)
  const block = []
  let consensusLine = ''
  if (!line) {
    return undefined
  }

  while (line) {
    if (line.startsWith(' ')) {
      consensusLine = line
    } else {
      block.push(line)
    }
    line = arr.next().value
  }
  const [start, end] = getSeqBounds(block[0])
  const fields = block.map(s => s.split(/\s+/))
  const ids = fields.map(s => s[0])
  const seqs = fields.map(s => s[2])
  let consensus = consensusLine.slice(start, end)

  // handle if the consensus trailing whitespace got trimmed
  const remainder = seqs[0].length - consensus.length
  if (remainder) {
    consensus += ' '.repeat(remainder)
  }
  return { ids, seqs, consensus }
}

export function parsePairwiseBlocks(arr: Iterator<string>) {
  let block
  const res = parsePairwiseBlock(arr)
  if (res !== undefined) {
    while ((block = parsePairwiseBlock(arr))) {
      for (let i = 0; i < block.seqs.length; i++) {
        res.seqs[i] += block.seqs[i]
      }
      res.consensus += block.consensus
    }
  }
  return res
}
