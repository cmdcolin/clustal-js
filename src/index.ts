import { parsePairwiseBlocks } from './pairwise'
import { parseBlocks, parseHeader, getFirstNonEmptyLine } from './util'

export function parseClustalIter(arr: Iterator<string>) {
  const line = getFirstNonEmptyLine(arr)
  if (!line) {
    throw new Error('Empty file received')
  }
  const header = parseHeader(line)

  const res = parseBlocks(arr)
  if (res === undefined) {
    throw new Error('No blocks parsed')
  }

  const alns = res.seqs.map((n, index) => ({ id: res.ids[index], seq: n }))
  const { consensus } = res
  if (consensus.length != alns[0].seq.length) {
    throw new Error(
      `Consensus length != sequence length. Con ${consensus.length} seq ${alns[0].seq.length}`,
    )
  }

  return { consensus, alns, header }
}

interface Row {
  id: string
  seq: string
}
export function parsePairwiseIter(arr: string) {
  const res = parsePairwiseBlocks(arr.split('\n')[Symbol.iterator]())
  if (res === undefined) {
    throw new Error('No blocks parsed')
  }

  const alns = res.seqs.map((n, index) => ({ id: res.ids[index], seq: n }))
  const { consensus } = res
  if (consensus.length != alns[0].seq.length) {
    throw new Error(
      `Consensus length != sequence length. Con ${consensus.length} seq ${alns[0].seq.length}`,
    )
  }
  if (alns.length !== 2) {
    throw new Error('More than two sequences in pairwise alignment')
  }
  return {
    consensus,
    alns: alns as [Row, Row],
  }
}

export function parse(contents: string) {
  const iter = contents.split('\n')[Symbol.iterator]()
  return parseClustalIter(iter)
}

export function parsePairwise(contents: string) {
  return parsePairwiseIter(
    contents
      .split('\n')
      .filter(f => !f.startsWith('#'))
      .join('\n'),
  )
}
