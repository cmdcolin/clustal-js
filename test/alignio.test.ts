import { parseString } from "../src";
// This is a truncated version of the example in Tests/cw02.aln
// Notice the inclusion of sequence numbers (right hand side)
const example1 = `CLUSTAL W (1.81) multiple sequence alignment


gi|4959044|gb|AAD34209.1|AF069      MENSDSNDKGSDQSAAQRRSQMDRLDREEAFYQFVNNLSEEDYRLMRDNN 50
gi|671626|emb|CAA85685.1|           ---------MSPQTETKASVGFKAGVKEYKLTYYTPEYETKDTDILAAFR 41
                                              * *: ::    :.   :*  :  :. : . :*  ::   .

gi|4959044|gb|AAD34209.1|AF069      LLGTPGESTEEELLRRLQQIKEGPPPQSPDENRAGESSDDVTNSDSIIDW 100
gi|671626|emb|CAA85685.1|           VTPQPG-----------------VPPEEAGAAVAAESSTGT--------- 65
                                    :   **                  **:...   *.*** ..

gi|4959044|gb|AAD34209.1|AF069      LNSVRQTGNTTRSRQRGNQSWRAVSRTNPNSGDFRFSLEINVNRNNGSQT 150
gi|671626|emb|CAA85685.1|           WTTVWTDGLTSLDRYKG-----RCYHIEPVPG------------------ 92
                                     .:*   * *: .* :*        : :* .*

gi|4959044|gb|AAD34209.1|AF069      SENESEPSTRRLSVENMESSSQRQMENSASESASARPSRAERNSTEAVTE 200
gi|671626|emb|CAA85685.1|           -EKDQCICYVAYPLDLFEEGSVTNMFTSIVGNVFGFKALRALRLEDLRIP 141
                                     *::.  .    .:: :*..*  :* .*   .. .  :    .  :

gi|4959044|gb|AAD34209.1|AF069      VPTTRAQRRA 210
gi|671626|emb|CAA85685.1|           VAYVKTFQGP 151
                                    *. .:: : .
`;
// This example is a truncated version of the dataset used here:
// http://virgil.ruc.dk/kurser/Sekvens/Treedraw.htm
// with the last record repeated twice (deliberate toture test)
const example2 = `CLUSTAL X (1.83) multiple sequence alignment


V_Harveyi_PATH                 --MKNWIKVAVAAIA--LSAA------------------TVQAATEVKVG
B_subtilis_YXEM                MKMKKWTVLVVAALLAVLSACG------------NGNSSSKEDDNVLHVG
B_subtilis_GlnH_homo_YCKK      MKKALLALFMVVSIAALAACGAGNDNQSKDNAKDGDLWASIKKKGVLTVG
YA80_HAEIN                     MKKLLFTTALLTGAIAFSTF-----------SHAGEIADRVEKTKTLLVG
FLIY_ECOLI                     MKLAHLGRQALMGVMAVALVAG---MSVKSFADEG-LLNKVKERGTLLVG
E_coli_GlnH                    --MKSVLKVSLAALTLAFAVS------------------SHAADKKLVVA
Deinococcus_radiodurans        -MKKSLLSLKLSGLLVPSVLALS--------LSACSSPSSTLNQGTLKIA
HISJ_E_COLI                    MKKLVLSLSLVLAFSSATAAF-------------------AAIPQNIRIG
HISJ_E_COLI                    MKKLVLSLSLVLAFSSATAAF-------------------AAIPQNIRIG
                                         : .                                 : :.

V_Harveyi_PATH                 MSGRYFPFTFVKQ--DKLQGFEVDMWDEIGKRNDYKIEYVTANFSGLFGL
B_subtilis_YXEM                ATGQSYPFAYKEN--GKLTGFDVEVMEAVAKKIDMKLDWKLLEFSGLMGE
B_subtilis_GlnH_homo_YCKK      TEGTYEPFTYHDKDTDKLTGYDVEVITEVAKRLGLKVDFKETQWGSMFAG
YA80_HAEIN                     TEGTYAPFTFHDK-SGKLTGFDVEVIRKVAEKLGLKVEFKETQWDAMYAG
FLIY_ECOLI                     LEGTYPPFSFQGD-DGKLTGFEVEFAQQLAKHLGVEASLKPTKWDGMLAS
E_coli_GlnH                    TDTAFVPFEFKQG--DKYVGFDVDLWAAIAKELKLDYELKPMDFSGIIPA
Deinococcus_radiodurans        MEGTYPPFTSKNE-QGELVGFDVDIAKAVAQKLNLKPEFVLTEWSGILAG
HISJ_E_COLI                    TDPTYAPFESKNS-QGELVGFDIDLAKELCKRINTQCTFVENPLDALIPS
HISJ_E_COLI                    TDPTYAPFESKNS-QGELVGFDIDLAKELCKRINTQCTFVENPLDALIPS
                                     **       .:  *::::.   : :.   .        ..:

V_Harveyi_PATH                 LETGRIDTISNQITMTDARKAKYLFADPYVVDG-AQI
B_subtilis_YXEM                LQTGKLDTISNQVAVTDERKETYNFTKPYAYAG-TQI
B_subtilis_GlnH_homo_YCKK      LNSKRFDVVANQVG-KTDREDKYDFSDKYTTSR-AVV
YA80_HAEIN                     LNAKRFDVIANQTNPSPERLKKYSFTTPYNYSG-GVI
FLIY_ECOLI                     LDSKRIDVVINQVTISDERKKKYDFSTPYTISGIQAL
E_coli_GlnH                    LQTKNVDLALAGITITDERKKAIDFSDGYYKSG-LLV
Deinococcus_radiodurans        LQANKYDVIVNQVGITPERQNSIGFSQPYAYSRPEII
HISJ_E_COLI                    LKAKKIDAIMSSLSITEKRQQEIAFTDKLYAADSRLV
HISJ_E_COLI                    LKAKKIDAIMSSLSITEKRQQEIAFTDKLYAADSRLV
                               *.: . *        .  *     *:          :

`;

const example3 = `CLUSTAL 2.0.9 multiple sequence alignment


Test1seq             ------------------------------------------------------------
AT3G20900.1-SEQ      ATGAACAAAGTAGCGAGGAAGAACAAAACATCAGGTGAACAAAAAAAAAACTCAATCCAC
AT3G20900.1-CDS      ------------------------------------------------------------


Test1seq             -----AGTTACAATAACTGACGAAGCTAAGTAGGCTACTAATTAACGTCATCAACCTAAT
AT3G20900.1-SEQ      ATCAAAGTTACAATAACTGACGAAGCTAAGTAGGCTAGAAATTAAAGTCATCAACCTAAT
AT3G20900.1-CDS      ------------------------------------------------------------


Test1seq             ACATAGCACTTAGAAAAAAGTGAAGTAAGAAAATATAAAATAATAAAAGGGTGGGTTATC
AT3G20900.1-SEQ      ACATAGCACTTAGAAAAAAGTGAAGCAAGAAAATATAAAATAATAAAAGGGTGGGTTATC
AT3G20900.1-CDS      ------------------------------------------------------------


Test1seq             AATTGATAGTGTAAATCATCGTATTCCGGTGATATACCCTACCACAAAAACTCAAACCGA
AT3G20900.1-SEQ      AATTGATAGTGTAAATCATAGTTGATTTTTGATATACCCTACCACAAAAACTCAAACCGA
AT3G20900.1-CDS      ------------------------------------------------------------


Test1seq             CTTGATTCAAATCATCTCAATAAATTAGCGCCAAAATAATGAAAAAAATAATAACAAACA
AT3G20900.1-SEQ      CTTGATTCAAATCATCTCAAAAAACAAGCGCCAAAATAATGAAAAAAATAATAACAAAAA
AT3G20900.1-CDS      ------------------------------------------------------------


Test1seq             AAAACAAACCAAAATAAGAAAAAACATTACGCAAAACATAATAATTTACTCTTCGTTATT
AT3G20900.1-SEQ      CAAACAAACCAAAATAAGAAAAAACATTACGCAAAACATAATAATTTACTCTTCGTTATT
AT3G20900.1-CDS      ------------------------------------------------------------


Test1seq             GTATTAACAAATCAAAGAGCTGAATTTTGATCACCTGCTAATACTACTTTCTGTATTGAT
AT3G20900.1-SEQ      GTATTAACAAATCAAAGAGATGAATTTTGATCACCTGCTAATACTACTTTCTGTATTGAT
AT3G20900.1-CDS      ------------------------------------------------------------


Test1seq             CCTATATCAACGTAAACAAAGATACTAATAATTAACTAAAAGTACGTTCATCGATCGTGT
AT3G20900.1-SEQ      CCTATATCAAAAAAAAAAAAGATACTAATAATTAACTAAAAGTACGTTCATCGATCGTGT
AT3G20900.1-CDS      ------------------------------------------------------ATGAAC
                                                                             *

Test1seq             TCGTTGACGAAGAAGAGCTCTATCTCCGGCGGAGCAAAGAAAACGATCTGTCTCCGTCGT
AT3G20900.1-SEQ      GCGTTGACGAAGAAGAGCTCTATCTCCGGCGGAGCAAAGAAAACGATCTGTCTCCGTCGT
AT3G20900.1-CDS      AAAGTAGCGAGGAAGAACAAAACATC------AGCAAAGAAAACGATCTGTCTCCGTCGT
                         *  *** ***** *   *  **      ****************************

Test1seq             AACACACGGTCGCTAGAGAAACTTTGCTTCTTCGGCGCCGGTGGACACGTCAGCATCTCC
AT3G20900.1-SEQ      AACACACAGTTTTTCGAGACCCTTTGCTTCTTCGGCGCCGGTGGACACGTCAGCATCTCC
AT3G20900.1-CDS      AACACACAGTTTTTCGAGACCCTTTGCTTCTTCGGCGCCGGTGGACACGTCAGCATCTCC
                     ******* **   * ****  ***************************************

Test1seq             GGTATCCTAGACTTCTTGGCTTTCGGGGTACAACAACCGCGTGGTGACGTCAGCACCGCT
AT3G20900.1-SEQ      GGTATCCTAGACTTCTTGGCTTTCGGGGTACAACAACCGCCTGGTGACGTCAGCACCGCT
AT3G20900.1-CDS      GGTATCCTAGACTTCTTGGCTTTCGGGGTACAACAACCGCCTGGTGACGTCAGCACCGCT
                     **************************************** *******************

Test1seq             GCTGGGGATGGAGAGGGAACAGAGTT-
AT3G20900.1-SEQ      GCTGGGGATGGAGAGGGAACAGAGTAG
AT3G20900.1-CDS      GCTGGGGATGGAGAGGGAACAGAGTAG
                     *************************
`;

const example4 = `Kalign (2.0) alignment in ClustalW format

Test1seq             GCTGGGGATGGAGAGGGAACAGAGTT-
AT3G20900.1-SEQ      GCTGGGGATGGAGAGGGAACAGAGTAG

`;

describe("alignio ported tests", () => {
  beforeEach(() => {});
  it("test one", () => {
    const alignment = parseString(example1);
    expect(alignment.length).toEqual(2);
    expect(alignment[0].id).toEqual("gi|4959044|gb|AAD34209.1|AF069");
    expect(alignment[1].id).toEqual("gi|671626|emb|CAA85685.1|");
    expect(alignment[0].seq).toEqual(
      "MENSDSNDKGSDQSAAQRRSQMDRLDREEAFYQFVNNLSEEDYRLMRDNN" +
        "LLGTPGESTEEELLRRLQQIKEGPPPQSPDENRAGESSDDVTNSDSIIDW" +
        "LNSVRQTGNTTRSRQRGNQSWRAVSRTNPNSGDFRFSLEINVNRNNGSQT" +
        "SENESEPSTRRLSVENMESSSQRQMENSASESASARPSRAERNSTEAVTE" +
        "VPTTRAQRRA"
    );
  });
  it("test two", () => {
    const alignment = parseString(example2);
    expect(alignment.length).toEqual(9);
    expect(alignment[-1].id).toEqual("HISJ_E_COLI");
    expect(alignment[-1]).toEqual(
      "MKKLVLSLSLVLAFSSATAAF-------------------AAIPQNIRIG" +
        "TDPTYAPFESKNS-QGELVGFDIDLAKELCKRINTQCTFVENPLDALIPS" +
        "LKAKKIDAIMSSLSITEKRQQEIAFTDKLYAADSRLV"
    );
  });
  it("test_empy", () => {
    expect(parseString("")).toMatchSnapshot()
  });

  it("test three", () => {
    const alignments = parseString(example3);
  });

  it("test kalign", () => {
    const alignments = parseString(example4);
    expect(alignments.length).toEqual(2);
  });
});
