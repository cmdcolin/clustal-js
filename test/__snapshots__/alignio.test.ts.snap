// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test three 1`] = `
{
  "alns": [
    {
      "id": "Test1seq",
      "seq": "-----------------------------------------------------------------AGTTACAATAACTGACGAAGCTAAGTAGGCTACTAATTAACGTCATCAACCTAATACATAGCACTTAGAAAAAAGTGAAGTAAGAAAATATAAAATAATAAAAGGGTGGGTTATCAATTGATAGTGTAAATCATCGTATTCCGGTGATATACCCTACCACAAAAACTCAAACCGACTTGATTCAAATCATCTCAATAAATTAGCGCCAAAATAATGAAAAAAATAATAACAAACAAAAACAAACCAAAATAAGAAAAAACATTACGCAAAACATAATAATTTACTCTTCGTTATTGTATTAACAAATCAAAGAGCTGAATTTTGATCACCTGCTAATACTACTTTCTGTATTGATCCTATATCAACGTAAACAAAGATACTAATAATTAACTAAAAGTACGTTCATCGATCGTGTTCGTTGACGAAGAAGAGCTCTATCTCCGGCGGAGCAAAGAAAACGATCTGTCTCCGTCGTAACACACGGTCGCTAGAGAAACTTTGCTTCTTCGGCGCCGGTGGACACGTCAGCATCTCCGGTATCCTAGACTTCTTGGCTTTCGGGGTACAACAACCGCGTGGTGACGTCAGCACCGCTGCTGGGGATGGAGAGGGAACAGAGTT-",
    },
    {
      "id": "AT3G20900.1-SEQ",
      "seq": "ATGAACAAAGTAGCGAGGAAGAACAAAACATCAGGTGAACAAAAAAAAAACTCAATCCACATCAAAGTTACAATAACTGACGAAGCTAAGTAGGCTAGAAATTAAAGTCATCAACCTAATACATAGCACTTAGAAAAAAGTGAAGCAAGAAAATATAAAATAATAAAAGGGTGGGTTATCAATTGATAGTGTAAATCATAGTTGATTTTTGATATACCCTACCACAAAAACTCAAACCGACTTGATTCAAATCATCTCAAAAAACAAGCGCCAAAATAATGAAAAAAATAATAACAAAAACAAACAAACCAAAATAAGAAAAAACATTACGCAAAACATAATAATTTACTCTTCGTTATTGTATTAACAAATCAAAGAGATGAATTTTGATCACCTGCTAATACTACTTTCTGTATTGATCCTATATCAAAAAAAAAAAAGATACTAATAATTAACTAAAAGTACGTTCATCGATCGTGTGCGTTGACGAAGAAGAGCTCTATCTCCGGCGGAGCAAAGAAAACGATCTGTCTCCGTCGTAACACACAGTTTTTCGAGACCCTTTGCTTCTTCGGCGCCGGTGGACACGTCAGCATCTCCGGTATCCTAGACTTCTTGGCTTTCGGGGTACAACAACCGCCTGGTGACGTCAGCACCGCTGCTGGGGATGGAGAGGGAACAGAGTAG",
    },
    {
      "id": "AT3G20900.1-CDS",
      "seq": "------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ATGAACAAAGTAGCGAGGAAGAACAAAACATC------AGCAAAGAAAACGATCTGTCTCCGTCGTAACACACAGTTTTTCGAGACCCTTTGCTTCTTCGGCGCCGGTGGACACGTCAGCATCTCCGGTATCCTAGACTTCTTGGCTTTCGGGGTACAACAACCGCCTGGTGACGTCAGCACCGCTGCTGGGGATGGAGAGGGAACAGAGTAG",
    },
  ],
  "consensus": "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            *       *  *** ***** *   *  **      *********************************** **   * ****  ******************************************************************************* ********************************************  ",
  "header": {
    "info": "CLUSTAL 2.0.9 multiple sequence alignment",
    "version": "2.0.9",
  },
}
`;
